src/oauth2/controller/authorize.ts

```diff
@@ -43,7 +43,7 @@ class AuthorizeController extends Controller {
     const responseType = ctx.query.response_type;
     const redirectUri = ctx.query.redirect_uri;
     const codeChallenge = ctx.query.code_challenge;
-    const codeChallengeMethod = ctx.query.code_challenge_method ? ctx.query.code_challenge_method : 'plain';
+    const codeChallengeMethod = ctx.query.code_challenge_method ? (<'S256'>ctx.query.code_challenge_method) : 'plain';
```

Comment:
This is effectively a type cast, if the code_challenge_method is garbage, it will be typed as `S256`.

Ideal I think would be a strict check. Error when it's not one of the 2 acceptable values.

Bit more verbose, but:

```suggestion
let codeChallengeMethod: CodeChallengeMethod = undefined;
switch(ctx.query.code_challenge_method) {
  case: 'S256':
  case 'plain' :
    codeChallengeMethod = ctx.query.code_challenge_method;
    break;
  default:
    throw new Error('...');
}
```

Reply:
I'm effectively doing the same check on line 33:
https://github.com/curveball/a12n-server/blob/29e05d92d4594f352cff53cc23ffab4150bff79a/src/oauth2/controller/authorize.ts#L33

Reply:
Maybe you can use an assertion function for easy re-use?

```
function assertIsChallengeMethod(input: string): asserts input is CodeChallengeMethod {

  if (input !== 'plain' && input !=='S256') {
    throw new Error('...');
  }
}
```

Used as such:

```
let codeChallengeMethod: CodeChallengeMethod | undefined;
if (ctx.query.code_challenge_method) {
   assertsIsChallengeMethod(ctx.query.code_challenge_method);
   codeChallengeMethod = ctx.query.code_challenge_method;
}
```

Reply:
An assertion function is 'trusted' by typescript to do the correct thing. Typescript will assume that you _will_ throw an error if the input was not correct.

After an assertion function is called, typescript will internally have narrowed down the `string` to `CodeChallengeMethod`.

Reply:
I created a type and went with the switch statement.